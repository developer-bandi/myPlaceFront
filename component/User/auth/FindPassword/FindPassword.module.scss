.mainBlock {
  width: 100%;
}
.title {
  text-align: center;
  font-weight: bold;
}
.subBlock {
  width: 600px;
  margin: auto;
  border-top: 2px solid black;
  border-bottom: 2px solid black;
  padding: 50px 30px 0px 30px;
  margin-bottom: 100px;
  @media screen and (max-width: 767px) {
    width: 80%;
    margin-bottom: 50px;
    padding: 50px 15px 0px 15px;
  }
}

.inputBlock {
  margin-bottom: 50px;
}

.input {
  width: 450px;
  height: 40px;
  border-bottom: 1px solid black;
  margin-right: 10px;
  @media screen and (max-width: 767px) {
    width: 75%;
  }
}

.button {
  width: 70px;
  height: 40px;
  border: 1px solid gray;
  border-radius: 5px;
  @media screen and (max-width: 767px) {
    width: 20%;
    font-size: 12px;
  }
}
