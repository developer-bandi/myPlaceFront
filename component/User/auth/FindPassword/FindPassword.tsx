import {RefObject} from "react";
import styles from "./FindPassword.module.scss";
import authStyles from "../../../../lib/styles/auth.module.scss";

interface FindPasswordProps {
  randomNumber:
    | {
        number: string;
        id: string;
      }
    | undefined;
  emailInputRef: RefObject<HTMLInputElement>;
  randomNumberInputRef: RefObject<HTMLInputElement>;
  sendMail: () => Promise<void>;
  checkAuthNum: () => void;
  authStatus: boolean;
  passwordRef: RefObject<HTMLInputElement>;
  passwordCheckRef: RefObject<HTMLInputElement>;
  changePassword: () => Promise<void>;
}

const FindPassword = ({
  randomNumber,
  emailInputRef,
  randomNumberInputRef,
  sendMail,
  checkAuthNum,
  authStatus,
  passwordRef,
  passwordCheckRef,
  changePassword,
}: FindPasswordProps) => {
  if (randomNumber === undefined && authStatus === false) {
    return (
      <main className={`${styles.mainBlock} ${styles.nomal}`}>
        <div className={authStyles.subBlock}>
          <h1 className={authStyles.title}>비밀번호 찾기</h1>
          <div className={authStyles.contentBlock}>
            <div className={authStyles.subTitle}>이메일</div>
            <input
              className={authStyles.input}
              ref={emailInputRef}
              placeholder="이메일을 입력하세요"
              key="email"
              onKeyPress={sendMail}
              data-testid="email"
            />
            <div className={authStyles.submitButtonBlock}>
              <button
                className={authStyles.submitButton}
                onClick={sendMail}
                data-testid="sendMail"
              >
                메일 전송
              </button>
            </div>
          </div>
        </div>
      </main>
    );
  } else if (randomNumber !== undefined && authStatus === false) {
    return (
      <main className={`${styles.mainBlock} ${styles.nomal}`}>
        <div className={authStyles.subBlock}>
          <h1 className={authStyles.title}>비밀번호 찾기</h1>
          <div className={authStyles.contentBlock}>
            <div className={authStyles.subTitle}>인증번호</div>
            <input
              className={authStyles.input}
              ref={randomNumberInputRef}
              placeholder="인증번호를 입력하세요"
              key="randomNumber"
              onKeyPress={checkAuthNum}
              data-testid="randomNumber"
            />
          </div>
          <div className={authStyles.submitButtonBlock}>
            <button
              className={authStyles.submitButton}
              onClick={checkAuthNum}
              data-testid="checkAuthNum"
            >
              인증번호 확인
            </button>
          </div>
        </div>
      </main>
    );
  } else {
    return (
      <main className={`${styles.mainBlock} ${styles.changePassword}`}>
        <div className={authStyles.subBlock}>
          <h1 className={authStyles.title}>비밀번호 찾기</h1>
          <div className={authStyles.contentBlock}>
            <div className={authStyles.subTitle}>비밀번호</div>
            <input
              className={authStyles.input}
              ref={passwordRef}
              placeholder="새 비밀번호를 입력하세요"
              type="password"
              key="newPassword"
              onKeyPress={changePassword}
              data-testid="newPassword"
            />
          </div>
          <div className={authStyles.contentBlock}>
            <div className={authStyles.subTitle}>비밀번호 확인</div>
            <input
              className={authStyles.input}
              ref={passwordCheckRef}
              placeholder="새 비밀번호를 한번더 입력하세요"
              type="password"
              key="newPasswordCheck"
              onKeyPress={changePassword}
              data-testid="newPasswordCheck"
            />
            <div className={authStyles.submitButtonBlock}>
              <button
                className={authStyles.submitButton}
                onClick={changePassword}
                data-testid="changePassword"
              >
                비밀번호 변경
              </button>
            </div>
          </div>
        </div>
      </main>
    );
  }
};

export default FindPassword;
