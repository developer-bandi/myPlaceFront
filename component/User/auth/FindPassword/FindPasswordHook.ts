import {useRouter} from "next/router";
import {useRef, useState} from "react";
import {axiosPostId, axiosPostPassword} from "../../../../lib/commonFn/api";
const CryptoJS = require("crypto-js");

const useFindPassword = () => {
  const emailInputRef = useRef<HTMLInputElement>(null);
  const randomNumberInputRef = useRef<HTMLInputElement>(null);
  const passwordRef = useRef<HTMLInputElement>(null);
  const passwordCheckRef = useRef<HTMLInputElement>(null);
  const router = useRouter();
  const [email, setEmail] = useState("");
  const [randomNumber, setRandomNumber] = useState<{
    number: string;
    id: string;
  }>();
  const [authStatus, setAuthStatus] = useState(false);

  const sendMail = async (e: {key?: string; type: string}) => {
    if (
      (e.type === "click" && e.key === undefined) ||
      (e.type === "keypress" && e.key === "Enter")
    ) {
      try {
        if (emailInputRef.current !== null) {
          const authResult = await axiosPostId(emailInputRef.current.value);
          if (authResult.status === 203) {
            alert("이메일이 존재하지 않습니다");
          } else {
            setEmail(emailInputRef.current.value);
            setRandomNumber({
              number: authResult.data.number,
              id: authResult.data.id,
            });
            setAuthStatus(false);
            alert("인증번호를 발송하였습니다");
          }
        }
      } catch (error) {
        alert("오류가 발생하였습니다");
      }
    }
  };

  const checkAuthNum = (e: {key?: string; type: string}) => {
    if (
      (e.type === "click" && e.key === undefined) ||
      (e.type === "keypress" && e.key === "Enter")
    ) {
      if (randomNumberInputRef.current !== null) {
        if (
          randomNumberInputRef.current.value ==
          CryptoJS.AES.decrypt(
            randomNumber?.number,
            process.env.NEXT_PUBLIC_PASSWORD_SECRET
          ).toString(CryptoJS.enc.Utf8)
        ) {
          alert("인증되었습니다");
          setAuthStatus(true);
        } else {
          alert("인증번호가 일치하지 않습니다");
        }
      }
    }
  };

  const changePassword = async (e: {key?: string; type: string}) => {
    if (
      (e.type === "click" && e.key === undefined) ||
      (e.type === "keypress" && e.key === "Enter")
    ) {
      if (
        passwordRef.current !== null &&
        passwordCheckRef.current !== null &&
        authStatus
      ) {
        if (passwordRef.current.value.length > 10) {
          alert("비밀번호 길이를 10자리 이하로 줄여주세요");
        } else if (
          passwordRef.current.value !== passwordCheckRef.current.value
        ) {
          alert("두 비밀번호가 일치하지 않습니다");
        } else {
          try {
            await axiosPostPassword(
              email,
              CryptoJS.AES.encrypt(
                JSON.stringify(passwordRef.current.value),
                process.env.NEXT_PUBLIC_PASSWORD_SECRET
              ).toString()
            );
            alert("변경되었습니다");
            router.push("/user/auth/signin");
          } catch (error) {
            alert("에러가 발생하였습니다");
          }
        }
      }
    }
  };

  return {
    randomNumber,
    emailInputRef,
    sendMail,
    randomNumberInputRef,
    checkAuthNum,
    authStatus,
    passwordRef,
    passwordCheckRef,
    changePassword,
    setRandomNumber,
    setAuthStatus,
  };
};

export default useFindPassword;
