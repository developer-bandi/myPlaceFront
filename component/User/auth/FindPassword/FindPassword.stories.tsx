import { ComponentStory, ComponentMeta } from "@storybook/react";
import FindPassword from "./FindPassword";
import { action } from "@storybook/addon-actions";

export default {
  title: "User/auth/FindPassword",
  component: FindPassword,
  parameters: {
    layout: "fullscreen",
  },
} as ComponentMeta<typeof FindPassword>;

const Template: ComponentStory<typeof FindPassword> = (arg) => {
  return <FindPassword {...arg} />;
};

export const inputEmail = Template.bind({});
inputEmail.args = {
  authStatus: false,
  sendMail: () =>
    new Promise((resolve) => {
      action("sendMail")();
      resolve();
    }),
};

export const inputRandomNumber = Template.bind({});
inputRandomNumber.args = {
  randomNumber: {
    number: "123456",
    id: "test",
  },
  authStatus: false,
  checkAuthNum: () =>
    new Promise((resolve) => {
      action("checkAuthNum")();
      resolve(1);
    }),
};

export const inputNewPassword = Template.bind({});
inputNewPassword.args = {
  randomNumber: {
    number: "123456",
    id: "test",
  },
  authStatus: true,
  changePassword: () =>
    new Promise((resolve) => {
      action("sendMail")();
      resolve();
    }),
};
