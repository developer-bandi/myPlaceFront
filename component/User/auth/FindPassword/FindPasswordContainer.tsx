import FindPassword from "./FindPassword";
import useFindPassword from "./FindPasswordHook";

const FindPasswordContainer = () => {
  const {
    randomNumber,
    emailInputRef,
    sendMail,
    randomNumberInputRef,
    checkAuthNum,
    authStatus,
    passwordRef,
    passwordCheckRef,
    changePassword,
  } = useFindPassword();
  return (
    <FindPassword
      randomNumber={randomNumber}
      emailInputRef={emailInputRef}
      sendMail={sendMail}
      randomNumberInputRef={randomNumberInputRef}
      checkAuthNum={checkAuthNum}
      authStatus={authStatus}
      passwordRef={passwordRef}
      passwordCheckRef={passwordCheckRef}
      changePassword={changePassword}
    ></FindPassword>
  );
};

export default FindPasswordContainer;
