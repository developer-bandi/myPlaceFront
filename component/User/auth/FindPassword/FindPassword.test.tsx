import {
  act,
  fireEvent,
  render,
  renderHook,
  screen,
} from "@testing-library/react";
import axios from "axios";
import newReact from "react";
import * as React from "react";
import {useRouter} from "next/router";
import useFindPassword from "./FindPasswordHook";
import FindPassword from "./FindPassword";
const CryptoJS = require("crypto-js");

jest.mock("axios");
const mockedAxios = axios as jest.Mocked<typeof axios>;

jest.mock("next/router", () => ({
  __esModule: true,
  useRouter: jest.fn(),
}));
const mockRouter = {
  push: jest.fn(),
};
(useRouter as jest.Mock).mockReturnValue(mockRouter);

const alertMock = jest.fn();
window.alert = alertMock;

describe("FindPassword Presentational 테스트", () => {
  it("이메일 입력단계", () => {
    const sendMailMock = jest.fn();
    const utils = render(
      <FindPassword
        randomNumber={undefined}
        emailInputRef={{current: null}}
        randomNumberInputRef={{current: null}}
        sendMail={sendMailMock}
        checkAuthNum={jest.fn()}
        authStatus={false}
        passwordRef={{current: null}}
        passwordCheckRef={{current: null}}
        changePassword={jest.fn()}
      />
    );
    expect(utils.container).toMatchSnapshot();
    fireEvent.click(screen.getByTestId("sendMail"));
    expect(sendMailMock).toBeCalled();
  });

  it("인증번호 입력단계", () => {
    const checkAuthNumMock = jest.fn();
    const utils = render(
      <FindPassword
        randomNumber={{id: "test", number: "test"}}
        emailInputRef={{current: null}}
        randomNumberInputRef={{current: null}}
        sendMail={jest.fn()}
        checkAuthNum={checkAuthNumMock}
        authStatus={false}
        passwordRef={{current: null}}
        passwordCheckRef={{current: null}}
        changePassword={jest.fn()}
      />
    );
    expect(utils.container).toMatchSnapshot();
    fireEvent.click(screen.getByTestId("checkAuthNum"));
    expect(checkAuthNumMock).toBeCalled();
  });

  it("비밀번호 입력단계", () => {
    const changePasswordMock = jest.fn();
    const utils = render(
      <FindPassword
        randomNumber={{id: "test", number: "test"}}
        emailInputRef={{current: null}}
        randomNumberInputRef={{current: null}}
        sendMail={jest.fn()}
        checkAuthNum={jest.fn()}
        authStatus={true}
        passwordRef={{current: null}}
        passwordCheckRef={{current: null}}
        changePassword={changePasswordMock}
      />
    );
    expect(utils.container).toMatchSnapshot();
    fireEvent.click(screen.getByTestId("changePassword"));
    expect(changePasswordMock).toBeCalled();
  });
});

describe("FindPassword Hook 테스트", () => {
  describe("sendMail 함수 테스트", () => {
    it("이메일이 존재하지 않는 경우", async () => {
      jest.spyOn(newReact, "useRef").mockReturnValue({current: {value: ""}});
      mockedAxios.post.mockImplementation(() =>
        Promise.resolve({
          status: 203,
          data: "아이디 중복",
        })
      );
      const {result} = renderHook(() => useFindPassword());
      await act(async () => {
        await result.current.sendMail();
      });
      expect(alertMock.mock.calls[0][0]).toBe("이메일이 존재하지 않습니다");
    });

    it("정상적으로 메일을 발송한 경우", async () => {
      jest.spyOn(newReact, "useRef").mockReturnValue({current: {value: ""}});
      mockedAxios.post.mockImplementation(() =>
        Promise.resolve({
          status: 200,
          data: {id: "testId", number: "testNumber"},
        })
      );
      const {result} = renderHook(() => useFindPassword());
      await act(async () => {
        await result.current.sendMail();
      });
      expect(result.current.randomNumber).toStrictEqual({
        id: "testId",
        number: "testNumber",
      });
      expect(result.current.authStatus).toBe(false);
      expect(alertMock.mock.calls[1][0]).toBe("인증번호를 발송하였습니다");
    });

    it("에러가 발생한 경우", async () => {
      jest.spyOn(newReact, "useRef").mockReturnValue({current: {value: ""}});
      mockedAxios.post.mockImplementation(() =>
        Promise.reject({
          status: 500,
          data: "error",
        })
      );
      const {result} = renderHook(() => useFindPassword());
      await act(async () => {
        await result.current.sendMail();
      });
      expect(alertMock.mock.calls[2][0]).toBe("오류가 발생하였습니다");
    });
  });

  describe("checkAuthNum 함수 테스트", () => {
    it("인증번호가 일치하는 경우", async () => {
      jest
        .spyOn(newReact, "useRef")
        .mockReturnValue({current: {value: "testNumber"}});
      jest
        .spyOn(CryptoJS.AES, "decrypt")
        .mockReturnValue({toString: () => "testNumber"});
      const {result} = renderHook(() => useFindPassword());
      act(() => {
        result.current.setRandomNumber({
          id: "testId",
          number: "testNumber",
        });
      });
      act(() => {
        result.current.checkAuthNum();
      });

      expect(alertMock.mock.calls[3][0]).toBe("인증되었습니다");
      expect(result.current.authStatus).toBe(true);
    });
    it("인증번호가 불일치하는 경우", async () => {
      jest
        .spyOn(newReact, "useRef")
        .mockReturnValue({current: {value: "testNumber"}});
      jest
        .spyOn(CryptoJS.AES, "decrypt")
        .mockReturnValue({toString: () => "notMatchtestNumber"});
      const {result} = renderHook(() => useFindPassword());
      act(() => {
        result.current.setRandomNumber({
          id: "testId",
          number: "testNumber",
        });
      });
      act(() => {
        result.current.checkAuthNum();
      });
      expect(alertMock.mock.calls[4][0]).toBe("인증번호가 일치하지 않습니다");
    });
  });

  describe("changePassword 함수 테스트", () => {
    it("비밀번호 길이가 조건에 맞지 않는 경우", async () => {
      jest
        .spyOn(newReact, "useRef")
        .mockReturnValueOnce({current: {value: "test@amail.com"}})
        .mockReturnValueOnce({current: {value: "123456"}})
        .mockReturnValueOnce({current: {value: "testPassword"}})
        .mockReturnValueOnce({current: {value: "testPassword"}})
        .mockReturnValueOnce({current: {value: "test@amail.com"}})
        .mockReturnValueOnce({current: {value: "123456"}})
        .mockReturnValueOnce({current: {value: "testPassword"}})
        .mockReturnValueOnce({current: {value: "testPassword"}});
      const {result} = renderHook(() => useFindPassword());
      act(() => {
        result.current.setAuthStatus(true);
        result.current.emailInputRef;
      });
      await act(async () => {
        await result.current.changePassword();
      });

      expect(alertMock.mock.calls[5][0]).toBe(
        "비밀번호 길이를 10자리 이하로 줄여주세요"
      );
    });
    it("두 비밀번호가 일치하지 않는 경우", async () => {
      jest
        .spyOn(newReact, "useRef")
        .mockReturnValueOnce({current: {value: "test@amail.com"}})
        .mockReturnValueOnce({current: {value: "123456"}})
        .mockReturnValueOnce({current: {value: "test"}})
        .mockReturnValueOnce({current: {value: "testPasswordnotMatch"}})
        .mockReturnValueOnce({current: {value: "test@amail.com"}})
        .mockReturnValueOnce({current: {value: "123456"}})
        .mockReturnValueOnce({current: {value: "test"}})
        .mockReturnValueOnce({current: {value: "testPasswordnotMatch"}});
      const {result} = renderHook(() => useFindPassword());
      act(() => {
        result.current.setAuthStatus(true);
      });
      await act(async () => {
        await result.current.changePassword();
      });
      expect(alertMock.mock.calls[6][0]).toBe(
        "두 비밀번호가 일치하지 않습니다"
      );
    });
    it("비밀번호가 변경된 경우", async () => {
      jest
        .spyOn(newReact, "useRef")
        .mockReturnValueOnce({current: {value: "test@amail.com"}})
        .mockReturnValueOnce({current: {value: "123456"}})
        .mockReturnValueOnce({current: {value: "test"}})
        .mockReturnValueOnce({current: {value: "test"}});
      jest
        .spyOn(CryptoJS.AES, "encrypt")
        .mockReturnValue({toString: () => "testPassword"});
      mockedAxios.post.mockImplementation(() =>
        Promise.resolve({
          status: 200,
          data: "test",
        })
      );
      const {result} = renderHook(() => useFindPassword());
      act(() => {
        result.current.setAuthStatus(true);
      });
      await act(async () => {
        await result.current.changePassword();
      });
      expect(alertMock.mock.calls[7][0]).toBe("변경되었습니다");
      expect(mockRouter.push).toBeCalledWith("/user/auth/signin");
    });
    it("에러가 발생한 경우", async () => {
      jest
        .spyOn(newReact, "useRef")
        .mockReturnValueOnce({current: {value: "test@amail.com"}})
        .mockReturnValueOnce({current: {value: "123456"}})
        .mockReturnValueOnce({current: {value: "test"}})
        .mockReturnValueOnce({current: {value: "test"}});
      mockedAxios.post.mockImplementation(() =>
        Promise.reject({
          status: 500,
          data: "test",
        })
      );
      const {result} = renderHook(() => useFindPassword());
      act(() => {
        result.current.setAuthStatus(true);
      });
      await act(async () => {
        await result.current.changePassword();
      });
      expect(alertMock.mock.calls[8][0]).toBe("에러가 발생하였습니다");
    });
  });
});
