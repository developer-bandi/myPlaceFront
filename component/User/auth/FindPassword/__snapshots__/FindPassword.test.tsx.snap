// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FindPassword Presentational 테스트 비밀번호 입력단계 1`] = `
<div>
  <main
    class="mainBlock changePassword"
  >
    <div
      class="subBlock"
    >
      <h1
        class="title"
      >
        비밀번호 찾기
      </h1>
      <div
        class="contentBlock"
      >
        <div
          class="subTitle"
        >
          비밀번호
        </div>
        <input
          class="input"
          data-testid="newPassword"
          placeholder="새 비밀번호를 입력하세요"
          type="password"
        />
      </div>
      <div
        class="contentBlock"
      >
        <div
          class="subTitle"
        >
          비밀번호 확인
        </div>
        <input
          class="input"
          data-testid="newPasswordCheck"
          placeholder="새 비밀번호를 한번더 입력하세요"
          type="password"
        />
        <div
          class="submitButtonBlock"
        >
          <button
            class="submitButton"
            data-testid="changePassword"
          >
            비밀번호 변경
          </button>
        </div>
      </div>
    </div>
  </main>
</div>
`;

exports[`FindPassword Presentational 테스트 이메일 입력단계 1`] = `
<div>
  <main
    class="mainBlock nomal"
  >
    <div
      class="subBlock"
    >
      <h1
        class="title"
      >
        비밀번호 찾기
      </h1>
      <div
        class="contentBlock"
      >
        <div
          class="subTitle"
        >
          이메일
        </div>
        <input
          class="input"
          data-testid="email"
          placeholder="이메일을 입력하세요"
        />
        <div
          class="submitButtonBlock"
        >
          <button
            class="submitButton"
            data-testid="sendMail"
          >
            메일 전송
          </button>
        </div>
      </div>
    </div>
  </main>
</div>
`;

exports[`FindPassword Presentational 테스트 인증번호 입력단계 1`] = `
<div>
  <main
    class="mainBlock nomal"
  >
    <div
      class="subBlock"
    >
      <h1
        class="title"
      >
        비밀번호 찾기
      </h1>
      <div
        class="contentBlock"
      >
        <div
          class="subTitle"
        >
          인증번호
        </div>
        <input
          class="input"
          data-testid="randomNumber"
          placeholder="인증번호를 입력하세요"
        />
      </div>
      <div
        class="submitButtonBlock"
      >
        <button
          class="submitButton"
          data-testid="checkAuthNum"
        >
          인증번호 확인
        </button>
      </div>
    </div>
  </main>
</div>
`;
